### PR TITLE
Large (💀) refactor of boundary conditions

### PR DESCRIPTION
**Pull request by almaeder**
This PR unifies the boundary conditions for the OBC and the Lyapunov problems.

This is mainly important for the memorization and the system size reduction.

The following needs to be done:
- [x] Before continuing, wait for #198 to progress
- [x] Remove warning from obc solver since the wrapper should warn
- [x] Adapt tests for the obc
- [x] Adapt quatrex to initialize the right classes
- [x] Change calls in the solver since now the residuals are returned as well.

Note: Example parameters needed to change:
There is always, after all methods, a single ref iteration to compute the residuals.
This matters not in practice, but the reference tests are very sensible since there are very few energy points.

### CODE REVIEW

**Comment by almaeder** [comment]
Currently, there is a minor issue: What should the behavior be when using QTBM since now not the injection vectors, but `phi_surface` is returned. Thus, `phi_surface` is computed with a `x_ii` with one ref iteration less than the returned `x_ii`.
This only materialized if `OBCSystem` which was intented since otherwise QTBM does not have a residual check.

Options:
- Accept that `phi_surface` is less accurate. (not good option)
- Not have a residual check in QTBM. (not good option)
- Change QTBM back to return the injection vectors and not `phi_surface`. Then eigenvalue scaling and `x_ii_a_ij`` needs to be done later.
- Only use `Spectral` in QTBM and add back in a residual check there.
- Building `phi_surface`  in `_get_residuals`, but then spectral needs to return injection vectors and corresponding eigenvalues.

**Comment by almaeder** [comment]
I believe the following is the best option:
> Only use `Spectral` in QTBM and add back in a residual check there.
The only question is which parameter should be used for the warning.
I believe it should be a separate keyword for qtbm.

**Comment by almaeder** [comment]
The contributions from this PR are:
- Unifying the memoizer between OBC and Lyapunov
- Restructuring how the system size reduction is done.
- Enabling future extension for system size reduction in the non-periodic dimension for OBC.
- Enabling future extension of returning the residual to be able to filter wrong obc points in the solvers.
- Smaller changes like removing OBC tests with two contact strings.

File: src/qttools/boundary_conditions/obc/spectral.py [hunk 1]
```diff
@@ -462,22 +469,19 @@ def __call__(
...
-            xp.linalg.norm(x_ii_ref - x_ii, axis=(-2, -1))
-            / xp.linalg.norm(x_ii_ref, axis=(-2, -1))
-        )
-        if recursion_error > self.warning_threshold:
-            warnings.warn(
-                f"High relative recursion error: {recursion_error:.2e}",
-                RuntimeWarning,
-            )
-
         # Calculate the injection vector and return it together with the boundary self-energy and the injected eigenvalues
```
Comment (maurodossena) [inline src/qttools/boundary_conditions/obc/spectral.py:487]: Here it's returning the surface Green's function and the injected eigenvectors

**Comment by almaeder** [comment]
> Currently, there is a minor issue: What should the behavior be when using QTBM since now not the injection vectors, but `phi_surface` is returned. Thus, `phi_surface` is computed with a `x_ii` with one ref iteration less than the returned `x_ii`. This only materialized if `OBCSystem` which was intented since otherwise QTBM does not have a residual check.
> 
> Options:
> 
>     * Accept that `phi_surface` is less accurate. (not good option)
> 
>     * Not have a residual check in QTBM. (not good option)
> 
>     * Change QTBM back to return the injection vectors and not `phi_surface`. Then eigenvalue scaling and `x_ii_a_ij`` needs to be done later.
> 
>     * Only use `Spectral` in QTBM and add back in a residual check there.
> 
>     * Building `phi_surface`  in `_get_residuals`, but then spectral needs to return injection vectors and corresponding eigenvalues.

@maurodossena @vetschn we need to discuss about QTBM again. How will we later separate and share the periodicity in the non-transport direction transformation? My initial thought was to handle it in the `contract`/`expand` of the OBC system, but then we had the discussion about the special case of QTBM.

Similarly, we need to discuss how to handle low rank approximation for QTBM, which does not need refinement iterations.

@vetschn is probably correct to separate the memoizer and the system size reduction.

**Comment by almaeder** [comment]
The system size reduction is now done by composition, and thus it can be reused later in both SCBA and QTBM.

File: src/qttools/boundary_conditions/obc/obc.py [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+        cache_compressor: None = None,
+        system_reducer: SystemReducer | None = None,
+        num_ref_iterations: int = 2,
+        relative_tol: float = 2e-1,
+        absolute_tol: float = 1e-6,
+        warning_threshold: float = 1e-1,
+        mode: str = "auto",
+        agreement_threshold: float = 0.999,
+    ) -> None:
+        """Initializes the lyapunov system."""
```
Comment (maurodossena) [inline src/qttools/boundary_conditions/obc/obc.py]: Is it only for lyapunov?

File: src/qttools/boundary_conditions/boundary_system.py [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+
+        # First, deflate the system
+        reduced_system = self.system_reducer.contract_system(boundary_system)
+
+        # memoize and cache on the reduced system
+        reduced_solution = self._memoized_solve(reduced_system, contact, **kwargs)
+
+        rel_residuals, abs_residuals, reduced_solution = self._get_residuals(
+            reduced_system, reduced_solution
+        )
```
Comment (maurodossena) [inline src/qttools/boundary_conditions/boundary_system.py:342]: The function `_get_residuals` is also refining the solution, which is unclear to me. I would rename it to explicity indicate that refining is also performed.

File: src/qttools/boundary_conditions/boundary_system.py [hunk 4]
```diff
@@ -0,0 +1,340 @@
...
+            if global_memoizing > self.agreement_threshold:
+                # If the result did not converge, recompute it from scratch.
+                return self._solve(boundary_system, contact, **kwargs)
+
+        for __ in range(self.num_ref_iterations - 2):
+            solution = self._fix_point_step(boundary_system, solution)
+
+        return solution
+
+    def __call__(
```
Comment (maurodossena) [inline src/qttools/boundary_conditions/boundary_system.py:303]: Will QTBM call this method in the future? At the moment it directly calls `spectral` right?

**Review by maurodossena — APPROVED** [review]
Looks good! One step further the unification of OBCs for WF and NEGF

File: src/qttools/boundary_conditions/boundary_system.py (hunk 4 above)
Comment (almaeder) [reply src/qttools/boundary_conditions/boundary_system.py:303]: QTBM will only call spectral in the future. Thus, the system size reduction is passed in and can be reused in both SCBA and QTBM.

File: src/qttools/boundary_conditions/boundary_system.py (hunk 3 above)
Comment (almaeder) [reply src/qttools/boundary_conditions/boundary_system.py:342]: Hmm, what would you suggest? The refinement is more a side product since it needs to be done to get the residuals.

File: src/qttools/boundary_conditions/obc/obc.py (hunk 2 above)
Comment (almaeder) [reply src/qttools/boundary_conditions/obc/obc.py]: :skull: not updated docstring. Thank you for catching.

File: examples/w90/carbon-nanotube/gw/quatrex_config.toml
Comment (vetschn) [inline examples/w90/carbon-nanotube/gw/quatrex_config.toml:1]: Why are the memoizers set to `"off"` in all examples? Not something we want to test?

File: src/qttools/boundary_conditions/lyapunov/lyapunov.py [hunk 5]
```diff
@@ -0,0 +1,497 @@
...
+            system_reducer=system_reducer,
+            num_ref_iterations=num_ref_iterations,
+            relative_tol=relative_tol,
+            absolute_tol=absolute_tol,
+            warning_threshold=warning_threshold,
+            mode=mode,
+            agreement_threshold=agreement_threshold,
+        )
+
+    def _fix_point_step(
```
Comment (vetschn) [inline src/qttools/boundary_conditions/lyapunov/lyapunov.py]: `_fixed_point_step` 😬

File: src/qttools/boundary_conditions/boundary_system.py [hunk 6]
```diff
@@ -0,0 +1,340 @@
...
+from qttools.boundary_conditions.system_reduction import IdentityReducer, SystemReducer
+from qttools.comm import comm
+
+
+class IdentityCompressor:
+    def compress(self, x):
+        return x.copy() if hasattr(x, "copy") else x
+
+    def decompress(self, x):
+        return x
```
Comment (vetschn) [inline src/qttools/boundary_conditions/boundary_system.py:45]: Missing docstrings, just a quick explanation.

File: src/qttools/boundary_conditions/boundary_system.py [hunk 7]
```diff
@@ -0,0 +1,340 @@
...
+        cache compression. If None, no compression is applied.
+    num_ref_iterations : int, optional
+        The number of fixed-point iterations to refine the solution. Default is 2.
+    relative_tol : float, optional
+        The relative tolerance for convergence. Default is 0.2.
+    absolute_tol : float, optional
+        The absolute tolerance for convergence. Default is 1e-6.
+    warning_threshold : float, optional
+        The threshold for issuing a warning about high residuals. Default is 0.1.
+    mode : str, optional
```
Comment (vetschn) [inline src/qttools/boundary_conditions/boundary_system.py]: Maybe rename this keyword to `memoization_mode`

File: src/qttools/boundary_conditions/boundary_system.py (hunk 3 above)
Comment (vetschn) [reply src/qttools/boundary_conditions/boundary_system.py:342]: Maybe just add a description to the docstring? Or do you think the implicit fixed-point step is really bad @maurodossena ?

File: src/qttools/boundary_conditions/system_reduction.py [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+
+
+class SystemReducer(ABC):
+
+    @abstractmethod
+    def contract_system(
+        self,
+        boundary_system: tuple[NDArray, ...],
+    ) -> tuple[NDArray, ...]:
+        """Contract the boundary system to a reduced system.
```
Comment (vetschn) [inline src/qttools/boundary_conditions/system_reduction.py:20]: I understand now what it is supposed to do but this needs just a little bit more explanation here. Along the lines of "The boundary systems can exhibit sparsity, which we can exploit to more efficiently compute solutions ..."

File: src/qttools/boundary_conditions/system_reduction.py [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+        """
+        ...
+
+
+class IdentityReducer(SystemReducer):
+    def contract_system(
+        self,
+        boundary_system: tuple[NDArray, ...],
+    ) -> tuple[NDArray, ...]:
+        """Contract the boundary system to a reduced system.
```
Comment (vetschn) [inline src/qttools/boundary_conditions/system_reduction.py]: Docstring needs to be updated

File: src/quatrex/coulomb_screening/solver.py [hunk 10]
```diff
@@ -194,7 +194,7 @@ def _compute_obc(self) -> None:
                     block_sections=self.block_sections,
                 )
 
-                x_00 = self.obc(a_ii=m_00, a_ij=m_01, a_ji=m_10, contact="left")
+                x_00, _, _ = self.obc((m_00, m_01, m_10), contact="left")
```
Comment (vetschn) [inline src/quatrex/coulomb_screening/solver.py]:
```suggestion
                x_00, *__ = self.obc((m_00, m_01, m_10), contact="left")
```

File: src/quatrex/coulomb_screening/solver.py [hunk 11]
```diff
@@ -228,7 +228,8 @@ def _compute_obc(self) -> None:
                 b_00 = x_00 @ m_10
                 q_00 = xp.stack((q_00_lesser, q_00_greater))
 
-                w_00_lesser, w_00_greater = self.lyapunov(b_00, q_00, "left")
+                w_00, _, _ = self.lyapunov((b_00, q_00), "left")
```
Comment (vetschn) [inline src/quatrex/coulomb_screening/solver.py]:
```suggestion
                w_00, *__ = self.lyapunov((b_00, q_00), "left")
```

File: src/quatrex/coulomb_screening/solver.py [hunk 12]
```diff
@@ -260,11 +261,13 @@ def _compute_obc(self) -> None:
                 m_nn = xp.flip(m_nn, axis=(-2, -1))
                 m_nm = xp.flip(m_nm, axis=(-2, -1))
                 m_mn = xp.flip(m_mn, axis=(-2, -1))
-                x_nn = self.obc(
+                x_nn, _, _ = self.obc(
```
Comment (vetschn) [inline src/quatrex/coulomb_screening/solver.py]:
```suggestion
                x_nn, *__ = self.obc(
```

File: src/quatrex/coulomb_screening/solver.py [hunk 13]
```diff
@@ -304,7 +307,8 @@ def _compute_obc(self) -> None:
 
                 q_nn = xp.stack((q_nn_lesser, q_nn_greater))
 
-                w_nn_lesser, w_nn_greater = self.lyapunov(b_nn, q_nn, "right")
+                w_nn, _, _ = self.lyapunov((b_nn, q_nn), "right")
```
Comment (vetschn) [inline src/quatrex/coulomb_screening/solver.py]:
```suggestion
                w_nn, *__ = self.lyapunov((b_nn, q_nn), "right")
```

File: src/quatrex/electron/solver.py [hunk 14]
```diff
@@ -317,10 +317,9 @@ def _compute_obc(self) -> None:
                 block_sections=self.block_sections,
             )
 
-            g_00 = self.obc(
-                a_ii=m_00 + s_00,
-                a_ij=m_01 + s_01,
-                a_ji=m_10 + s_10,
+            # TODO: use residuals to filter "bad" energies
+            g_00, _, _ = self.obc(
```
Comment (vetschn) [inline src/quatrex/electron/solver.py]:
```suggestion
            g_00, *__ = self.obc(
```

File: src/quatrex/electron/solver.py [hunk 15]
```diff
@@ -356,11 +355,13 @@ def _compute_obc(self) -> None:
             m_nn = xp.flip(m_nn, axis=(-2, -1))
             m_nm = xp.flip(m_nm, axis=(-2, -1))
             m_mn = xp.flip(m_mn, axis=(-2, -1))
-            g_nn = self.obc(
+            g_nn, _, _ = self.obc(
```
Comment (vetschn) [inline src/quatrex/electron/solver.py]:
```suggestion
            g_nn, *__ = self.obc(
```

File: tests/qttools/boundary_conditions/lyapunov/test_lyapunov.py [hunk 16]
```diff
@@ -0,0 +1,86 @@
...
+
+    lyapunov_system_reducer = LyapunovSystemReducer(reduce_sparsity=reduce_sparsity)
+
+    lyapunov_system = LyapunovSystem(
+        lyapunov_solver,
+        system_reducer=lyapunov_system_reducer,
+        mode="off",
+    )
+
+    a, q, _, _ = inputs
```
Comment (vetschn) [inline tests/qttools/boundary_conditions/lyapunov/test_lyapunov.py]:
```suggestion
    a, q, *__ = inputs
```

File: tests/qttools/boundary_conditions/lyapunov/test_lyapunov.py [hunk 17]
```diff
@@ -0,0 +1,86 @@
...
+
+    x, _, _ = lyapunov_system((a, q), "contact")
```
Comment (vetschn) [inline tests/qttools/boundary_conditions/lyapunov/test_lyapunov.py]:
```suggestion
    x, *__ = lyapunov_system((a, q), "contact")
```

File: tests/qttools/boundary_conditions/lyapunov/test_lyapunov.py [hunk 18]
```diff
@@ -0,0 +1,86 @@
...
+):
+    lyapunov_system_reducer = LyapunovSystemReducer(reduce_sparsity=reduce_sparsity)
+
+    lyapunov_system = LyapunovSystem(
+        lyapunov_solver,
+        system_reducer=lyapunov_system_reducer,
+        mode="off",
+    )
+
+    a, q, _, _ = inputs
```
Comment (vetschn) [inline tests/qttools/boundary_conditions/lyapunov/test_lyapunov.py]:
```suggestion
    a, q, *__ = inputs
```

File: tests/qttools/boundary_conditions/lyapunov/test_lyapunov.py [hunk 19]
```diff
@@ -0,0 +1,86 @@
...
+
+    a[:] = 0
+
+    x, _, _ = lyapunov_system((a, q), "contact")
```
Comment (vetschn) [inline tests/qttools/boundary_conditions/lyapunov/test_lyapunov.py]:
```suggestion
    x, *__ = lyapunov_system((a, q), "contact")
```

File: tests/qttools/boundary_conditions/lyapunov/test_lyapunov.py [hunk 20]
```diff
@@ -0,0 +1,86 @@
...
+    assert xp.allclose(x, a @ x @ a.conj().swapaxes(-1, -2) + q)
+
+    # Add a little noise to the input matrices.
+    a = a + 1e-3 * a
+    q = q + 1e-3 * q
+
+    a[..., row_slice, col_slice] = 0
+    a[..., row_slice, col_slice] = 0
+
+    x, _, _ = lyapunov_system((a, q), contact="contact")
```
Comment (vetschn) [inline tests/qttools/boundary_conditions/lyapunov/test_lyapunov.py]:
```suggestion
    x, *__ = lyapunov_system((a, q), contact="contact")
```

File: tests/qttools/boundary_conditions/obc/test_sancho_rubio.py [hunk 21]
```diff
@@ -30,21 +30,21 @@ def test_max_iterations(a_xx: tuple[NDArray, ...]):
...
+    x_ii, _, _ = obc_system((a_ii, a_ij, a_ji), contact="contact")
     assert xp.allclose(x_ii, xp.linalg.inv(a_ii - a_ji @ x_ii @ a_ij), atol=1e-5)
 
     # Add a little noise to the input matrices.
     a_ji = a_ji * (1 + 1e-6)
     a_ii = a_ii * (1 + 1e-6)
     a_ij = a_ij * (1 + 1e-6)
 
-    x_ii = spectral(a_ii=a_ii, a_ij=a_ij, a_ji=a_ji, contact=contact)
+    x_ii, _, _ = obc_system((a_ii, a_ij, a_ji), contact="contact")
```
Comment (vetschn) [inline tests/qttools/boundary_conditions/obc/test_sancho_rubio.py]:
```suggestion
    x_ii, *__ = obc_system((a_ii, a_ij, a_ji), contact="contact")
```

File: tests/qttools/boundary_conditions/obc/test_spectral.py [hunk 22]
```diff
@@ -156,7 +156,7 @@ def test_memoizer(
         (a_ji_hat, a_ii_hat, a_ij_hat), block_sections
     )
 
-    x_ii = spectral(a_ii=a_ii, a_ij=a_ij, a_ji=a_ji, contact=contact)
+    x_ii, _, _ = obc_system((a_ii, a_ij, a_ji), contact="contact")
```
Comment (vetschn) [inline tests/qttools/boundary_conditions/obc/test_spectral.py]:
```suggestion
    x_ii, *__ = obc_system((a_ii, a_ij, a_ji), contact="contact")
```

File: tests/qttools/boundary_conditions/obc/test_spectral.py [hunk 23]
```diff
@@ -167,7 +167,7 @@ def test_memoizer(
         < 5e-3
     )
 
-    x_ii = spectral(a_ii=a_ii_hat, a_ij=a_ij_hat, a_ji=a_ji_hat, contact=contact)
+    x_ii, _, _ = obc_system((a_ii_hat, a_ij_hat, a_ji_hat), contact="contact")
```
Comment (vetschn) [inline tests/qttools/boundary_conditions/obc/test_spectral.py]:
```suggestion
    x_ii, *__ = obc_system((a_ii_hat, a_ij_hat, a_ji_hat), contact="contact")
```

**Review by vetschn — COMMENTED** [review]
Usual small nits, overall very nice work

File: examples/w90/carbon-nanotube/gw/quatrex_config.toml
Comment (almaeder) [reply examples/w90/carbon-nanotube/gw/quatrex_config.toml:1]: In the integration tests, the memoizer will not trigger. Also, now the behaviour of the number of reference iterations is different: 
- Off: 
  - the ones in the used solver + 1 (to get residuals)
- Auto:
  - First iteration: the ones in the used solver + 1 (to get residuals)
  - Second+ iteration: 1 to test for memoization +the ones in the used solver + 1 (to get residuals)

Thus, with auto, the tests currently fail since there is one more ref iteration starting with the second iteration.

File: src/qttools/boundary_conditions/lyapunov/lyapunov.py (hunk 5 above)
Comment (almaeder) [reply src/qttools/boundary_conditions/lyapunov/lyapunov.py]: 💀

File: src/qttools/boundary_conditions/lyapunov/lyapunov.py [hunk 24]
```diff
@@ -0,0 +1,576 @@
...
+                "The system reduction information is missing.\n"
+                + "Make sure to call '_contract_system' before contracting the system."
+            )
+
+        a, q = boundary_system
+        a_hat = a[..., self.cols_reduced_system, self.cols_reduced_system]
+        q_hat = q[..., self.cols_reduced_system, self.cols_reduced_system]
+        return a_hat, q_hat
+
+    def __contract_system_zero_rows(
```
Comment (vetschn) [inline src/qttools/boundary_conditions/lyapunov/lyapunov.py]: Is there a real point to using the __dunder name mangling for these methods? I would change to just single leading underscore.

**Review by vetschn — APPROVED** [review]
Very nice changes splitting up the different tasks involved in solving for boundary conditions into smaller coherent parts.

All tests passing on alps.
